lib/Bindings/Python/SupportModule.cpp

```diff
@@ -0,0 +1,69 @@
+//===- SupportModule.cpp - Support API pybind module ----------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "SupportModule.h"
+
+#include "circt/Support/BackedgeBuilder.h"
+#include "circt/Support/LLVM.h"
+
+#include "mlir-c/IR.h"
+#include "mlir/CAPI/IR.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/Dialect.h"
+
+#include "PybindUtils.h"
+#include "pybind11/pybind11.h"
+namespace py = pybind11;
+
+using namespace circt;
+
+//===----------------------------------------------------------------------===//
+// Classes that translate from something Pybind11 understands to MLIR C++.
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+class PyBackedge {
+public:
+  PyBackedge(Backedge backedge) : backedge(backedge) {}
+
+  MlirValue value() { return wrap(backedge); }
+
+  void setValue(MlirValue newValue) { backedge.setValue(unwrap(newValue)); }
+
+private:
+  Backedge backedge;
```

Comment:
I'm not sure this is "safe". I've run into weird stuff with MLIR IR types when not using the C API, which I've resolved by storing the MLIR C type instead of the C++ one. This member stores a MlirValue, so you might see some weirdness here.

I also don't think that using the C++ backedge builder adds much value but it does add a fair amount of complexity (and potential for breakage which we don't really understand)... It should be relatively simple to implement this functionality in pure Python and not require any upstream changes, right? I've also thought about doing exactly this in this way, but now that I see it...

Reply:
So far, this seems to work. This member is technically storing an mlir::Value, but I know what you meant. If there is an issue with this approach, and we want to go this route, I can make this a proper C API that only stores the MlirValue struct.

I did consider just replicating this functionality in Python, and it is definitely doable. The duplication didn't sit right with me, so I whipped up this module. Happy to do it just in Python. I think we will end up at least needing to extend what is in this current module at the Python level to make the error reporting nicer. At that point, it may be simpler to just do it all in Python.

I think there still needs to be an upstream change either way. You can see here a commented out line where at one point I wasn't even using the backedge to set the value: https://github.com/llvm/circt/pull/940/files#diff-45ef9b344a85b7561366d5f9457f7de8579b64d97164cf4e4adc21462b11ef1bR102. Something like that is nice, except that API doesn't exist upstream. Whether we add that API, use the backedge to set the value, or do something else, the bottom line is there is no way right now to pass PyValue <> MlirValue back and forth from Python to C++ pybind modules, and I think anything we do here is going to require that.

Reply:
Yep, you're right. And I hear you about the duplication.

Reply:
Upstreamed the support needed to do this in Python.